workers/datasets_based/src/datasets_based/worker_factory.py

```diff
@@ -34,38 +39,61 @@ def _create_worker(self, job_info: JobInfo) -> Worker:
         pass
 
 
+@dataclass
 class WorkerFactory(BaseWorkerFactory):
-    def __init__(self, app_config: AppConfig) -> None:
-        self.app_config = app_config
+    app_config: AppConfig
+    processing_graph: ProcessingGraph
+    hf_datasets_cache: Path
 
     def _create_worker(self, job_info: JobInfo) -> Worker:
         job_type = job_info["type"]
         try:
-            processing_step = self.app_config.processing_graph.graph.get_step_by_job_type(job_type)
+            processing_step = self.processing_graph.get_step_by_job_type(job_type)
         except ValueError as e:
             raise ValueError(
                 f"Unsupported job type: '{job_type}'. The job types declared in the processing graph are:"
-                f" {[step.job_type for step in self.app_config.processing_graph.graph.steps.values()]}"
+                f" {[step.job_type for step in self.processing_graph.steps.values()]}"
             ) from e
         if job_type == ConfigNamesWorker.get_job_type():
-            return ConfigNamesWorker(job_info=job_info, app_config=self.app_config, processing_step=processing_step)
+            return ConfigNamesWorker(
+                job_info=job_info,
+                app_config=self.app_config,
+                processing_step=processing_step,
+                hf_datasets_cache=self.hf_datasets_cache,
+            )
         if job_type == SplitNamesWorker.get_job_type():
-            return SplitNamesWorker(job_info=job_info, app_config=self.app_config, processing_step=processing_step)
+            return SplitNamesWorker(
+                job_info=job_info,
+                app_config=self.app_config,
+                processing_step=processing_step,
+                hf_datasets_cache=self.hf_datasets_cache,
+            )
         if job_type == SplitsWorker.get_job_type():
-            return SplitsWorker(job_info=job_info, app_config=self.app_config, processing_step=processing_step)
-        if job_type == FirstRowsWorker.get_job_type():
-            first_rows_config = FirstRowsConfig.from_env()
-            return FirstRowsWorker(
+            return SplitsWorker(
                 job_info=job_info,
                 app_config=self.app_config,
                 processing_step=processing_step,
-                first_rows_config=first_rows_config,
+                hf_datasets_cache=self.hf_datasets_cache,
             )
+        if job_type == FirstRowsWorker.get_job_type():
+            first_rows_config = FirstRowsConfig.from_env()
+            with AssetsDirectoryResource(
+                init_storage_directory=first_rows_config.assets.storage_directory
+            ) as resource:
```

Comment:
Hmmm. Maybe the resource is not the directory as itself, but the access to the storage.

The idea is to ensure from the start that we have access to all the external resources (log, disk, databases), that's what I implement in a follow-up PR:

https://github.com/huggingface/datasets-server/pull/792/files#diff-7ba05e0a51a2335dfeefffff4bea2c3c17f1d5d0c60fdc4cbb90e049a86aebfeR49

Would it be better if I change the name of the resource so that it is clear that we will not delete the directory afterward?

Reply:
Done in https://github.com/huggingface/datasets-server/pull/784/commits/d99d0e468a7e5d5f03ba227762cf8b71768d8eea